opacclient/tests/src/main/java/de/geeksfactory/opacclient/apis/WebOpacAtTest.java

```diff
@@ -0,0 +1,45 @@
+package de.geeksfactory.opacclient.apis;
+
+import org.json.JSONObject;
+import org.junit.Test;
+
+import java.util.Collections;
+import java.util.Objects;
+
+import de.geeksfactory.opacclient.objects.Library;
+import de.geeksfactory.opacclient.searchfields.SearchQuery;
+
+public class WebOpacAtTest {
+
+    WebOpacAt getApi() {
+        WebOpacAt api = new WebOpacAt();
+        Library lib = new Library();
+        lib.setData(new JSONObject(Collections.singletonMap("baseurl", "http://www.stbinnsbruck.web-opac.at")));
+        api.init(lib);
+        return api;
+    }
+
+    @Test
+    public void testParseTotalResults() {
+        assert WebOpacAt.parseTotalResults("Treffer pro Seite  •  1 - 10 von 46 Treffern") == 46;
+        assert WebOpacAt.parseTotalResults("Hits per page  •  1 - 10 of 46") == 46;
+        assert WebOpacAt.parseTotalResults("Sayfa başına sonuç  •  46 sonuçtan 1 - 10") == 46;
+    }
+
+    @Test
+    public void testParseCopyReturn() throws Exception {
+        assert Objects.equals(WebOpacAt.parseCopyReturn("1 (voraussichtl. bis 31.07.2015)"), "31.07.2015");
+
+    }
+
+    @Test
+    public void testSearchFields() throws Exception {
+        getApi().getSearchFields();
```

Comment:
We already use `getSearchFields` in the generic Unit tests and test if there is at least a search field for free search or search by title, so I think an individual test for `getSearchFields` would not be necessary.
The only thing I could imagine a generic `testEmptyQuery` test doing would be to check if either a result list is returned or an `OpacErrorException` is thrown (which can show a user-understandable error message).

I also like the idea of having a different type of unit tests that only test if the APIs work in general instead of in every single library. Those could even be run on the CI server to make sure we don't destroy anything important when adjusting things in the APIs. But I understand that we should not rely on the OPAC of a specific library being online for that, too.
So the three options I see for doing this without relying on a library OPAC would be 
1. Have a demo instance of every OPAC we support under our control for running these kinds of tests, which is probably quite unrealistic for most commercial systems and doable, but probably hard to set up for open source systems like VuFind or Koha.
2. Build a mock server for every OPAC which is run by the unit tests themselves (using Jetty?) and would behave similar to the library servers, but could return prepared, static HTML responses (e.g. the same result list for every search query). Then we could run generic unit tests for all APIs on those
3. Similar to the previous one but a little simpler: Build specific unit tests for every API that could call functions like `parse_search`, `parse_result` (BTW @raphaelm : Is there a reason why you didn't use camelCase for these names?) etc. using prepared HTML. This would be "one level higher" compared to the two specific unit tests `testParseTotalResults` and `testParseCopyReturn` here.

All of these options would even allow us to test account related features, which is something we cannot do in the generic unit tests.
The latter two options would be doable in theory, but probably a great deal of work to build (and to maintain when adding new OPACs or a different version of the same OPAC).